src/main/java/com/scalar/db/server/DistributedTransactionService.java

```diff
@@ -0,0 +1,292 @@
+            Mutation mutation = mutations.get(0);
+            if (mutation instanceof Put) {
+              transaction.put((Put) mutation);
+            } else {
+              transaction.delete((Delete) mutation);
+            }
+          } else {
+            boolean hasPut = false;
+            boolean hasDelete = false;
+            for (Mutation mutation : mutations) {
+              if (mutation instanceof Put) {
+                hasPut = true;
+              } else {
+                hasDelete = true;
+              }
+            }
+            boolean mixed = hasPut & hasDelete;
+
+            if (mixed) {
+              transaction.mutate(mutations);
+            } else if (hasPut) {
+              transaction.put(mutations.stream().map(m -> (Put) m).collect(Collectors.toList()));
+            } else {
+              transaction.delete(
+                  mutations.stream().map(m -> (Delete) m).collect(Collectors.toList()));
+            }
+          }
+          responseObserver.onNext(Empty.newBuilder().build());
+          responseObserver.onCompleted();
+        },
+        responseObserver);
+  }
+
+  @Override
+  public void commit(CommitRequest request, StreamObserver<Empty> responseObserver) {
+    execute(
+        () -> {
+          DistributedTransaction transaction = getTransaction(request.getTransactionId());
+          try {
+            transaction.commit();
+          } finally {
+            activeTransactions.remove(transaction.getId());
+          }
+          responseObserver.onNext(Empty.newBuilder().build());
+          responseObserver.onCompleted();
+        },
+        responseObserver);
+  }
+
+  @Override
+  public void abort(AbortRequest request, StreamObserver<Empty> responseObserver) {
+    execute(
+        () -> {
+          DistributedTransaction transaction = getTransaction(request.getTransactionId());
+          try {
+            transaction.abort();
+          } finally {
+            activeTransactions.remove(transaction.getId());
+          }
+          responseObserver.onNext(Empty.newBuilder().build());
+          responseObserver.onCompleted();
+        },
+        responseObserver);
+  }
+
+  @Override
+  public void getState(
+      GetTransactionStateRequest request,
+      StreamObserver<GetTransactionStateResponse> responseObserver) {
+    execute(
+        () -> {
+          TransactionState state = manager.getState(request.getTransactionId());
+          responseObserver.onNext(
+              GetTransactionStateResponse.newBuilder()
+                  .setState(ProtoUtil.toTransactionState(state))
+                  .build());
+          responseObserver.onCompleted();
+        },
+        responseObserver);
+  }
+
+  private DistributedTransaction getTransaction(String transactionId) throws TransactionException {
+    ActiveTransaction activeTransaction = activeTransactions.get(transactionId);
+    if (activeTransaction == null) {
+      throw new TransactionException(
```

Comment:
A question and a comment.
I think this case happens mostly in a case where different connections are used in a transaction, which only happens when a connection is closed for some reason after the transaction is started.
First of all, is this right?

If so, should the exception be treated separately?
I feel it's not an `INTERNAL` error.
What do you think?

Reply:
I think this case could happen due to the following reasons:
1. A client specifies a wrong transaction ID
2. A client specifies a transaction ID that associates with a transaction that's already closed 
3. A transaction associated with the specified transaction ID is already expired by the `transactionExpirationThread`

An `INVALID_ARGUMENT` error suits the 1st and 2nd cases, but I think they are rare cases. And an `INTERNAL` error suits the 3rd case, and I think it most likely happens. So I decided to throw an `INTERNAL` error for this case. What do you think?